packetbeat/protos/amqp/amqp.go

```diff
@@ -0,0 +1,883 @@
+			message:  &AmqpMessage{Ts: pkt.Ts},
+		}
+	} else {
+		// concatenate databytes
+		priv.Data[dir].data = append(priv.Data[dir].data, pkt.Payload...)
+		if len(priv.Data[dir].data) > tcp.TCP_MAX_DATA_IN_STREAM {
+			logp.Debug("amqp", "Stream data too large, dropping TCP stream")
+			priv.Data[dir] = nil
+			return priv
+		}
+	}
+
+	stream := priv.Data[dir]
+
+	for len(stream.data) > 0 {
+		if stream.message == nil {
+			stream.message = &AmqpMessage{Ts: pkt.Ts}
+		}
+
+		ok, complete := amqp.amqpMessageParser(priv.Data[dir])
+		if !ok {
+			// drop this tcp stream. Will retry parsing with the next
+			// segment in it
+			priv.Data[dir] = nil
+			return priv
+		}
+		if complete {
+			amqp.handleAmqp(stream.message, tcptuple, dir)
+		} else {
+			// wait for more data
+			break
+		}
+	}
+	return priv
+}
+
+func (amqp *Amqp) GapInStream(tcptuple *common.TcpTuple, dir uint8,
+	nbytes int, private protos.ProtocolData) (priv protos.ProtocolData, drop bool) {
+	//TODO
+	return private, true
+}
+
+func (amqp *Amqp) ReceivedFin(tcptuple *common.TcpTuple, dir uint8,
+	private protos.ProtocolData) protos.ProtocolData {
+	//TODO
+	return private
+}
+
+func (amqp *Amqp) amqpMessageParser(s *AmqpStream) (ok bool, complete bool) {
+	for s.parseOffset < len(s.data) {
+
+		if len(s.data[s.parseOffset:]) < 8 {
+			logp.Warn("AMQP message smaller than a frame, discarding it")
+			return false, false
+		}
+
+		yes, version := isProtocolHeader(s.data[s.parseOffset:])
+		if yes {
+			logp.Debug("amqp", "Client header detected, version %d.%d.%d",
+				version[0], version[1], version[2])
+			s.parseOffset += 8
+			return true, false
+		}
+
+		f, err := readFrameHeader(s.data[s.parseOffset:])
+		if err {
+			//incomplete frame
+			return false, false
+		}
+
+		switch f.Type {
+		case METHOD:
+			ok, complete = amqp.decodeMethodFrame(s, f.content)
+		case HEADER:
+			ok = amqp.decodeHeaderFrame(s, f.content)
+		case BODY:
+			ok, complete = s.decodeBodyFrame(f.content)
+		case HEARTBEAT:
+			logp.Debug("amqpdetailed", "Heartbeat frame received")
+		default:
+			logp.Warn("Received unknown AMQP frame")
+			return false, false
+		}
+
+		// cast should be safe because f.size should not be bigger than tcp.TCP_MAX_DATA_IN_STREAM
+		s.parseOffset += 8 + int(f.size)
+		if !ok {
+			return false, false
+		} else if complete {
+			return true, true
+		}
+	}
+	return ok, complete
+}
+
+func (stream *AmqpStream) PrepareForNewMessage() {
+	stream.message = nil
+}
+
+func isProtocolHeader(data []byte) (isHeader bool, version string) {
+	if (string(data[:4]) == "AMQP") && data[4] == 0 {
+		return true, string(data[5:8])
+	}
+	return false, ""
+}
+
+//func to read a frame header and check if it is valid and complete
+func readFrameHeader(data []byte) (ret *AmqpFrame, err bool) {
+	var frame AmqpFrame
+
+	frame.size = binary.BigEndian.Uint32(data[3:7])
+	if len(data) < int(frame.size)+8 {
+		logp.Warn("Frame shorter than declared size, discarding it")
+		return nil, true
```

Comment:
frame being shorter really an error? When using tcp, is can be very well the case and frame will be continued with next TCP packet.